src/zeroband/infer.py

```diff
@@ -281,22 +282,114 @@ def inference(config: InferenceConfig):
             tokenize=True,
         )
 
-        # Convert BatchEncoding to TokensPrompt objects
-        token_prompts = [TokensPrompt(prompt_token_ids=input_ids) for input_ids in tokenized_prompts]
+        generate_start_time = time.time()
+        if config.contexts:
+            # Convert tokenized prompts to TokensPrompt objects (prompt_id -> TokensPrompt)
+            unordered_token_prompts: dict[int, TokensPrompt] = {
+                i: TokensPrompt(prompt_token_ids=prompt_token_ids) for i, prompt_token_ids in enumerate(tokenized_prompts)
+            }
+            # Save chunked request outputs (prompt_id -> RequestOutput)
+            unordered_request_outputs: dict[int, RequestOutput] = {}
+            unordered_proofs: dict[int, bytes] = defaultdict(bytes)
+
+            max_model_len = llm.llm_engine.model_config.max_model_len
+            assert max(config.contexts) <= max_model_len, "The final context should be the max model length"
+            num_finished_sequences = 0
+            for i, context in enumerate(config.contexts):
+                # Auto-compute the max batch size for the current context
+                logger.info(f"Running at context {context}")
+                local_max_batch_size = compute_max_batch_size(llm, max_model_len=context)
+                if node:
+                    max_batch_size = int(
+                        all_reduce(node, torch.tensor(local_max_batch_size), config=config.parallel.pp, op=torch.min).item()
+                    )
+                else:
+                    max_batch_size = local_max_batch_size
```

Comment:
this case actually isnt needed right? we would ever run with contexts without pp?